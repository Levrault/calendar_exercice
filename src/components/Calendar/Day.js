import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CalendarButton from '../Button/CalendarButton';
import EventFormPopUp from '../Event/EventFormPopUp';
import styles from './Day.css';

class Day extends Component {
  /**
  * @contructor
  * @param {object} props
  */
  constructor (props) {
    super(props);
    this.state = {
      selected: false
    };
  }

  /**
   * active day when clicked
   */
  handleClick = (event) => {
    this.setState({
      selected: !this.state.selected
    });
  }

  /**
  * Render
  */
  render () {
    const { selected } = this.state;
    const { value } = this.props;
    return (
      <div className={styles.day}>
        <EventFormPopUp active={selected} />
        <CalendarButton onClick={this.handleClick} active={selected}>
          {value}
        </CalendarButton>
      </div>
    );
  }
}

Day.propTypes = {
  value: PropTypes.number.isRequired
};

export default Day;
