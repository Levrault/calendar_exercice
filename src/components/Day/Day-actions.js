export const DAY_SELECTED = 'DAY_SELECTED';

/**
 * When a day is selected
 * @param {string} selectedDay
 * @returns {object}
 */
export const onSelected = (selectedDay) => (
  {
    type: DAY_SELECTED,
    payload: {
      selected: selectedDay
    }
  }
);

