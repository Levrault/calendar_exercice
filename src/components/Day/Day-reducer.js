import {
  DAY_SELECTED
} from './Day-actions';

const initialState = {
  selected: ''
};

/**
 * Profile from reducer
 * @param {object} state
 * @param {object} action
 * @returns {object}
 */
export const dayReducer = (state = initialState, action) => {
  switch (action.type) {
  case DAY_SELECTED:
    return {
      ...state,
      ...action.payload
    };
  default:
    return state;
  }
};

