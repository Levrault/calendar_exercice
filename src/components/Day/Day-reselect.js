
import { createSelector } from 'reselect';
import get from 'lodash/get';

/**
 * Create selector
 * @param {object} state
 * @returns {function}
 */
const daySelector = state => get(state, 'day');

export const daySelectedSelector = createSelector(
  daySelector,
  day => get(day, 'selected', '')
);
