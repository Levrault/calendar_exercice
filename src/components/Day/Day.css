.day {
  display: flex;
  flex: 1;
  max-width: 32px;
  position: relative;
}
