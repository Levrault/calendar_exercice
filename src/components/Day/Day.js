import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import CalendarButtonConnected from '../Button/CalendarButtonConnected';
import EventFormConnected from '../Event/EventFormConnected';
import EventShowConnected from '../Event/EventShowConnected';
import styles from './Day.css';

/**
 * Display/manage interaction of a day
 * @class Day
 * @param {string} date
 * @param {number} value
 * @param {function} onSelected
 * @param {bool} isSelected
 * @param {string} monthId
 * @param {object} [event]
 */
class Day extends PureComponent {
  /**
   * active/inactive day when clicked
   */
  handleClick = () => {
    const { onSelected, date, selectedDate, isSelected } = this.props;
    if (selectedDate !== date) {
      onSelected(date);
    } else if (isSelected) {
      onSelected('');
    }
  }

  /**
  * Render
  * @returns {node}
  */
  render () {
    const { value, isSelected, monthId, event } = this.props;
    const editMode = !!(event);

    return (
      <div className={styles.day}>
        {isSelected && !editMode &&
          <EventFormConnected
            onCancel={this.handleClick}
            monthId={monthId}
            day={value} />
        }

        {isSelected && editMode &&
          <EventShowConnected
            onCancel={this.handleClick}
            {...event} />
        }

        <CalendarButtonConnected
          onClick={this.handleClick}
          active={isSelected}
          event={event}
          editMode={editMode}>
          {value}
        </CalendarButtonConnected>
      </div>
    );
  }
}

Day.propTypes = {
  date: PropTypes.string.isRequired,
  value: PropTypes.number.isRequired,
  onSelected: PropTypes.func.isRequired,
  selectedDate: PropTypes.string.isRequired,
  isSelected: PropTypes.bool.isRequired,
  monthId: PropTypes.string.isRequired,
  event: PropTypes.shape({
    monthId: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    color: PropTypes.string.isRequired,
    day: PropTypes.number.isRequired,
    id: PropTypes.string.isRequired
  })
};

export default Day;
