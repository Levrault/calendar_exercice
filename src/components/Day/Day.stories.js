import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, number, text, boolean } from '@storybook/addon-knobs/react';
import { withMarkdownNotes } from '@storybook/addon-notes';
import DayWrapper from '../../tools/storybook/DayWrapper';
import Provider from '../../tools/storybook/Provider';
import Day from './Day';

storiesOf('Day', module)
  .addDecorator(withKnobs)
  .addDecorator(story => <Provider story={story()} />)
  .addDecorator(DayWrapper)
  .add('default', withMarkdownNotes(`
    # Day

    Is used create/show/delete an event
    `)(() => (
    <Day
      date={text('date', '1-1')}
      value={number('value', 1)}
      onSelected={() => {}}
      selectedDate={text('selectedDate', '')}
      isSelected={boolean('isSelected', false)}
      monthId="a4e8a03dbe54ca7d"
    />
  )))
  .add('selected', () => (
    <Day
      date={text('date', '1-1')}
      value={number('value', 1)}
      onSelected={() => {}}
      selectedDate={text('selectedDate', '1-1')}
      isSelected={boolean('isSelected', true)}
      monthId="a4e8a03dbe54ca7d"
    />
  ))
  .add('show', () => (
    <Day
      date={text('date', '1-1')}
      value={number('value', 1)}
      onSelected={() => {}}
      selectedDate={text('selectedDate', '1-1')}
      isSelected={boolean('isSelected', true)}
      monthId="a4e8a03dbe54ca7d"
      event={{
        monthId: 'a4e8a03dbe54ca7d',
        name: 'Event',
        color: '#EEEEEE',
        day: 1,
        id: '05af70ab81cc3a40'
      }}
    />
  ));
