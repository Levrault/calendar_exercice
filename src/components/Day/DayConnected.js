import { connect } from 'react-redux';
import Day from './Day';
import { onSelected } from './Day-actions';
import { daySelectedSelector } from './Day-reselect';

/**
 * @param {object} state
 * @param {object} ownProps
 * @returns {object}
 */
const mapStateToProps = (state, ownProps) => {
  const date = daySelectedSelector(state);
  return {
    selectedDate: date,
    isSelected: date === ownProps.date
  };
};

export default connect(mapStateToProps, { onSelected })(Day);
