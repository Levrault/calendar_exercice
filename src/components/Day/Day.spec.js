import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import getFakeStore from '../../tools/test/fakeStore';
import { shallow } from '../../tools/test/enzyme';
import Day from './Day';

const store = getFakeStore();

const mock = {
  date: '1-1',
  value: 1,
  onSelected: jest.fn(),
  selectedDate: '1-1',
  isSelected: false,
  monthId: 'a4e8a03dbe54ca7d',
  event: undefined
};

const mockEvents = {
  monthId: '0b54aaf1844478e4',
  name: 'hello',
  color: '#EEEEEE',
  fontColor: '#212121',
  day: 8,
  id: '05af70ab81cc3a40'
};

describe('Day component', () => {
  it('should render without problem', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Day
          {...mock}
        />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with an event', () => {
    const newMock = { ...mock, isSelected: true, event: mockEvents };

    const component = renderer.create(
      <Provider store={store}>
        <Day
          {...newMock}
        />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with event form', () => {
    const newMock = { ...mock, isSelected: true };

    const component = renderer.create(
      <Provider store={store}>
        <Day
          {...newMock}
        />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call onSelected on handleClick', () => {
    const component = shallow(
      <Day
        {...mock}
        selectedDate="2-1"
      />
    );

    component.instance().handleClick();
    expect(component.instance().props.onSelected).toHaveBeenCalled();
  });

  it('should call onSelected with empty value when selectedDate is different from date and isSelected is true', () => {
    const newMock = { ...mock, isSelected: true };
    const component = shallow(
      <Day
        {...newMock}
      />
    );

    component.instance().handleClick();
    expect(component.instance().props.onSelected).toHaveBeenCalled();
  });

  it('should not call onSelected selectedDate is different from date and isSelected is false', () => {
    const component = shallow(
      <Day
        {...mock}
      />
    );

    component.instance().handleClick();
    expect(component.instance().props.onSelected).toHaveBeenCalled();
  });
});
