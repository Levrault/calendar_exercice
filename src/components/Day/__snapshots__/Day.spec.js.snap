// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day component should render with an event 1`] = `
<div>
  <div
    style={
      Object {
        "backgroundColor": "#EEEEEE",
      }
    }
  >
    <div
      style={
        Object {
          "color": "#212121",
        }
      }
    >
      hello
    </div>
    <div>
      <button
        className=""
        onClick={[Function]}
        style={
          Object {
            "color": "#212121",
          }
        }
        type="button"
      >
        <i
          className="material-icons"
        >
          clear
        </i>
      </button>
      <button
        className=""
        onClick={[Function]}
        style={
          Object {
            "color": "#212121",
          }
        }
        type="button"
      >
        <i
          className="material-icons"
        >
          delete
        </i>
      </button>
    </div>
  </div>
  <button
    className="undefined"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#EEEEEE",
        "color": "#212121",
      }
    }
  >
    1
  </button>
</div>
`;

exports[`Day component should render with event form 1`] = `
<div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <div>
        <div>
          <input
            name="event"
            onChange={[Function]}
            required={true}
            style={
              Object {
                "borderBottom": "2px solid undefined",
              }
            }
            type="text"
            value=""
          />
          <label
            htmlFor="nodejs"
          >
            Rappel
          </label>
        </div>
      </div>
      <ul
        className=""
      >
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#FF5252",
              }
            }
            type="button"
            value="#FF5252"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#FF4081",
              }
            }
            type="button"
            value="#FF4081"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#E040FB",
              }
            }
            type="button"
            value="#E040FB"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#FF6E40",
              }
            }
            type="button"
            value="#FF6E40"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#448AFF",
              }
            }
            type="button"
            value="#448AFF"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#64FFDA",
              }
            }
            type="button"
            value="#64FFDA"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#B2FF59",
              }
            }
            type="button"
            value="#B2FF59"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#CDDC39",
              }
            }
            type="button"
            value="#CDDC39"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#795548",
              }
            }
            type="button"
            value="#795548"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#546E7A",
              }
            }
            type="button"
            value="#546E7A"
          />
        </li>
        <li>
          <button
            className=""
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#0097A7",
              }
            }
            type="button"
            value="#0097A7"
          />
        </li>
        <li>
          <button
            className="undefined"
            onClick={[Function]}
            style={
              Object {
                "backgroundColor": "#EEEEEE",
              }
            }
            type="button"
            value="#EEEEEE"
          />
        </li>
      </ul>
      <div>
        <button
          className=""
          onClick={[Function]}
          type="button"
        >
          <i
            className="material-icons"
          >
            clear
          </i>
        </button>
        <button
          className=""
          type="submit"
        >
          <i
            className="material-icons"
          >
            add
          </i>
        </button>
        <button
          className=""
          onClick={[Function]}
          type="button"
        >
          <i
            className="material-icons"
          >
            color_lens
          </i>
        </button>
      </div>
    </form>
  </div>
  <button
    className="undefined"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#EEEEEE",
        "color": "#212121",
      }
    }
  >
    1
  </button>
</div>
`;

exports[`Day component should render without problem 1`] = `
<div>
  <button
    className=""
    onClick={[Function]}
    style={Object {}}
  >
    1
  </button>
</div>
`;
